### DISCUSSION TITLE
Job with dynamic matrix crashes if matrix contains zero elements. How to resolve?

### DISCUSSION THREAD

**Discussion by cagross** [community/community#27096] [Actions]
<p>Hi all.  I have a GitHub action with two jobs.  The first job returns an array, which I’ve been able to pass to a second job which then uses that array as its matrix.  But the first job can sometimes return an array with zero length, and when it does, the second job crashes with the error:</p>
<p><code>Error when evaluating 'strategy' for job 'Job-2'. (Line: 52, Col: 11): Matrix vector 'cfg' does not contain any values</code></p>
<p>How can I resolve this?  Ideally, if job-1 returns an array with zero elements, the second job would not run at all.</p>
<p>I’ve seen <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepscontinue-on-error" rel="noopener nofollow ugc">this continue-on-error flag</a>, but I’m not sure exactly how to use it. In my case, the crash doesn’t seem to occur on any specific step–it occurs when the job is trying to evaluate the matrix.</p>
<hr>
<p><strong>My Code (myfile.yml)</strong></p>
<pre><code>name: GitHub Actions Demo
on: [push]
jobs:
  Job-1:
    outputs:
      matrix: ${{ steps.hello.outputs.matrix }}
    runs-on: ubuntu-latest
    steps:
      - run: echo "🎉 The job was automatically triggered by a ${{ github.event_name }} event."
      - name: Checkout
        uses: actions/checkout@v2.3.4
        with:
          fetch-depth: 0
      - name: Test JS action
        id: hello
        uses: myaccount/myrepo@master # This executes a JavaScript file which sets an array as its output.
Job-2:
    needs: Job-1
    runs-on: ubuntu-latest
    strategy:
      matrix:
        cfg:
          - ${{fromJson(needs.Job-1.outputs.matrix)}}
    #line 52 of the file.
    steps:
      - run: echo "🎉 SHA ${{ matrix.cfg.offSha }}."</code></pre>

**Comment by mickeygousset** [comment, marked as answer]
<p>Try putting some sort of condition if statement on Job 2, where you do something like:</p>
<p>if: ${{ needs.Job-1.outputs.matrix != “” }}</p>
<p>something like that.</p>

**Comment by cagross** [comment]
<p>OK perfect, thanks–that works!  I had found a way to set a condition on a particular step in a job, but didn’t know how to put a conditional on an entire job, which is exactly what I needed.</p>

**Comment by mickeygousset** [comment]
<p>Can you share what the if statement looks like?</p>

**Comment by cagross** [comment]
<p>It’s essentially what you gave me :slight_smile:</p>
<p><code>if: ${{ needs.Job-1.outputs.matrix != '' }}</code></p>
<p>At the same time though, I had to modify the previous job to return an empty string if it would otherwise return an empty array.  Previously it would return this string: <code>[]</code></p>

**Reply by dhanakane-vc** [reply]
I ran into a similar issue. I was tyring to evaluate the output of a prior job which returned an empty JSON array `[]` in certain cases. Using an `if` statement along the lines of `if ${{ needs.job1.outputs.foo != '[] }}` and `if needs.job1.outputs.foo != []` would not evaluate to false when `needs.job1.outputs.foo` was an empty array, which meant Github Actions kept throwing an error on the matrix value being empty.

This is not ideal, as convention states that if the input for my previous job is an array, then the output when the array is empty should be an empty array and not a string. It would be great to know how to properly handle this case.

**Reply by stijnveenman** [reply]
This saved me quite a bit of debugging, as I was testing the `Array` object against a string. One way I worked around this, while not having to change the outputs of a previous job is converting it back to a string when doing the check. `if: ${{ join(needs.job1.outputs.foo) != '' }}`. Join will concat items (with optional seperator). So if the array is empty this will create an empty string

**Comment by PoisonPhang** [comment]
For those looking to prevent builds with empty matrices in the form of empty JSON array's, a conditional in the following form worked for me:
```yaml
    if: ${{ needs.some-job.outputs.builds != '[]' }}
```
Where `builds` is the matrix in question.
Note the single quotes around the brackets.

**Reply by PoisonPhang** [reply]
cc @dhanakane-vc
